PF-27 Add .NET client for Samples

DougSchmidt-AI opened the pull request:
@jessemcdowell-AI Here's a first cut of a .NET Samples client, based on the methods exposed in the Swagger JSON. Let's review it in person when you're back from vacation and up to speed.

It builds upon some of the work & patterns that @DavidMills-AI started with the SamplesConnector project.

I've been able to use the client to upload attachments, import data, and make other requests.

I didn't include the hacked code generator that will regenerate the `ServiceModel.cs` file from a live Swagger file. My reasoning is that the Samples Swagger configuration needs a bit of love (like unique operationIds) to be a bit more standard. Once that work is done, most of the hacks in the code generator will go away, and we can commit that tool as well.

Commit:
PF-27 First cut of Samples client for .NET

1) Use the internal code generator to create ServiceModel.cs from the Samples Swagger 2.0 json spec.

2) Add a basic ISamplesClient to provide a similar API experience from .NET as the AQTS client.

Commit:
PF-27 Added Samples-specific exception classes and file upload support

Commit:
PF-27 Added tests for FileUploader class

Confirmed both import-service and regular attachment content formatting.

Commit:
PF-27 Added support for anonymous POST requests to Samples

There are a few API POST requests that don't fully follow RequestDTO/ResponseDTO pattern.
These requests accept anonymous lists of inputs, and return a ResponseDTO.

So we expose the InvokeWebServiceMethod() methods on the interface, so that the SamplesConnector can execute the lower-level JsonServiceClient methods with a correct SamplesContext.

This avoids polluting the longer-term .NET SDK interface with requests that are more brittle.

So we need to expose the less wrist-friendly Verb<TResponse>(string relativeOrAbsoluteUri, object request) methods until the Samples APIs are improved.

jessemcdowell-AI commented on src/Aquarius.Client/Samples/Client/FileUploader.cs:
```diff
@@ -0,0 +1,110 @@
+                AddFormDataContent(content, new StringContent(uploadedFilename), "qqfilename");
+                AddFormDataContent(content, new StringContent(Guid.NewGuid().ToString("D")), "qquuid");
+                AddFormDataContent(content, new StringContent(contentBytes.Length.ToString(CultureInfo.InvariantCulture)), "qqtotalfilesize");
+            });
```
This is pretty gross. Perhaps we should do something on the Samples API so public consumers can use more sensible names.

TheRealDougSchmidt replied:
Yup. Browser F12 was my friend here.

Uploading attachments (or even importing results) required lots of reverse engineering on my part in order to push a file successfully across the API.

For comparison, the AQTS endpoints used to receive file upload requests do not require any of the FineUploader multi-part content fields, so they work equally well when called directly from a .NET client or from a Javascript page using the FineUploader control.

jessemcdowell-AI commented on src/Aquarius.Client/Samples/Client/SamplesClient.cs:
```diff
@@ -0,0 +1,233 @@
+
+                count += responseDto.DomainObjects.Count;
+
+                if (count >= totalCount || !responseDto.DomainObjects.Any())
```
Depending on the source of the list, the total count could change between requests. This isn't a very safe mechanism to find the end of the list anyway.

I'd suggest testing if either the last request returned no objects, or no cursor to determine if you're at the end of the list.

TheRealDougSchmidt replied:
Good catch. I'll make that logic change to the lazy results iterator.

jessemcdowell-AI reviewed (commented):
Looks like a pretty good first attempt to me. I have only one minor comment.

Commit:
PF-27 Code review feedback

Improved the robustness of the lazy results evaluation, to account for changing TotalCounts or empty Cursor values in the responses.
